crates/rpc/rpc/src/eth/api/fees.rs

```diff
@@ -185,4 +187,92 @@ where
             reward: Some(rewards),
         })
     }
+
+    /// Populate the fee history cache when new blocks are added to the chain
+    ///
+    /// This listens for any new blocks and updates the fee history cache
+    async fn populate_fee_history_cache_on_new_block<St>(
+        &mut self,
+        reward_percentiles: Option<Vec<f64>>,
+        mut events: St,
+    ) where
+        St: Stream<Item = CanonStateNotification> + Unpin + 'static,
+    {
+        while let Some(event) = events.next().await {
+            match event {
+                CanonStateNotification::Reorg { old: _, new } => {
+                    let (new_blocks, _) = new.inner();
+                    match self.add_cache_for_block(&new_blocks.tip(), &reward_percentiles).await {
+                        Ok(_) => {}
+                        Err(e) => {
+                            warn!("Error adding block to fee history cache: {:?}", e);
+                        }
+                    }
+                }
+                CanonStateNotification::Commit { new } => {
+                    let (new_blocks, _) = new.inner();
+                    match self.add_cache_for_block(&new_blocks.tip(), &reward_percentiles).await {
+                        Ok(_) => {}
+                        Err(e) => {
+                            warn!("Error adding block to fee history cache: {:?}", e);
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+
+    /// Update fee history cache on new block
+    async fn add_cache_for_block(
+        &mut self,
+        block: &SealedBlockWithSenders,
+        reward_percentiles: &Option<Vec<f64>>,
+    ) -> Result<(), InvalidTransactionError> {
```

Comment:
is this entirely new code?

I assumed, that we are able to reuse a lot from the existing logic. is this not the case?

Reply:
It's the same caching logic used while reporting the fee history for the given blocks. I will refactor the code to avoid repetition.